### PR TITLE
Consolidate user role resolution for API keys

### PR DESCRIPTION
**Pull request by n1v0lg**
This refactor extracts the user role resolution logic for API keys from
`ApiKeyGenerator`. It plugs the shared resolver class into API key
creation and update handling. It also removes `ApiKeyGenerator` since
the class is now trivial. A new REST base handler for API key-related
REST actions ensures that the API key service is enabled _before_ we
perform role resolution, which was the only other responsibility left
to `ApiKeyGenerator`.

Relates:
https://github.com/elastic/elasticsearch/pull/88186/files#r910515818

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/apikey/TransportGrantApiKeyAction.java [hunk 1]
```diff
@@ -30,7 +30,7 @@
  */
 public final class TransportGrantApiKeyAction extends TransportGrantAction<GrantApiKeyRequest, CreateApiKeyResponse> {
 
-    private final ApiKeyGenerator generator;
+    private final ApiKeyManager apiKeyManager;
```
Comment (n1v0lg) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/apikey/TransportGrantApiKeyAction.java:33]: Decidedly not in love with `Manager`. The other alternative I have is `ApiKeyWriter` -- it's a little more specific which is good but `Writer` has a lot of `XContent` connotation which is bad.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/ApiKeyUserRoleDescriptorResolver.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import org.elasticsearch.xcontent.NamedXContentRegistry;
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.core.security.authc.Subject;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+import org.elasticsearch.xpack.core.security.authz.support.DLSRoleQueryValidator;
+import org.elasticsearch.xpack.security.authz.store.CompositeRolesStore;
+
+import java.util.Set;
+
+public class ApiKeyUserRoleDescriptorResolver {
```
Comment (n1v0lg) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/ApiKeyUserRoleDescriptorResolver.java:21]: I can also add a test for `ApiKeyUpdateHandler` however it's very simple and already exercised by our integration tests.

**Review by ywangd — COMMENTED** [review]
The essence of `ApiKeyGenerator` is really just about resolving the owner user's role descriptors. Once this part is refactored out to its own class, there is no good reason to still keep this class around. I'd suggest we simply remove it and let the handler classes (`TransportCreateApiKeyAction` and `TransportGrantApiKeyAction` etc) to be the true handlers. It will be a bit bigger change. But that's OK for a refactoring PR.

**Review by ywangd — APPROVED** [review]
LGTM

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/rest/action/apikey/ApiKeyBaseRestHandlerTests.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            protected RestChannelConsumer innerPrepareRequest(RestRequest request, NodeClient client) {
+                return channel -> {
+                    if (consumerCalled.compareAndSet(false, true) == false) {
+                        fail("consumerCalled was not false");
+                    }
+                };
+            }
+        };
+        final var fakeRestRequest = new FakeRestRequest();
+        final var fakeRestChannel = new FakeRestChannel(fakeRestRequest, randomBoolean(), serviceEnabled ? 0 : 1);
```
[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/rest/action/apikey/ApiKeyBaseRestHandlerTests.java:58]: For sure, good catch!

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/apikey/TransportGrantApiKeyAction.java [hunk 4]
```diff
@@ -71,14 +73,28 @@ public TransportGrantApiKeyAction(
...
+                            authentication,
+                            request.getApiKeyRequest(),
+                            roleDescriptors,
+                            listener
+                        ),
+                        listener::onFailure
+                    )
+                )
+            )
         );
```
[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/apikey/TransportGrantApiKeyAction.java:98]: Nice! I'll track this in a GH issue. Should be able to get to bigger tasks and if we don't it's a good low hanging fruit PR for a new future joiner.

**Pull request merged**